File: unpruuf/relay-android/app/src/main/java/com/nexonai/unpruuf/relay/core/BlobStore.kt
package com.nexonai.unpruuf.relay.core

import android.content.ContentValues
import android.content.Context
import android.database.DatabaseUtils
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

/**
 * Blind store-and-forward queue, addressed only by the sender's rotating wire tag — same model
 * as unpruuf/server/src/store/blobStore.ts, reimplemented with Android's built-in SQLite
 * (no Room/SQLCipher needed: unlike the main unpruuf app's contacts DB, everything stored here
 * is already Double-Ratchet-encrypted ciphertext the relay is architecturally blind to, so
 * at-rest encryption of the queue itself adds no real confidentiality — matching the Node
 * version's own plain-SQLite choice).
 */
class BlobStore(context: Context) : SQLiteOpenHelper(context, "relay.db", null, 1) {

    // Backs /v1/fetchMany's optional long-poll (see RelayHttpServer) — a caller thread can block
    // on several tags at once instead of the server blindly re-polling every few seconds. Purely
    // a process-local "wake up sooner" optimization, same as the Node relay's EventEmitter
    // equivalent: a missed wakeup just means the immediate re-check fetchMany always does after
    // waiting (see waitForAny's caller) still finds the blob on the next regular poll.
    // NanoHTTPD handles each request on its own thread and blocks it synchronously, so a
    // CountDownLatch per waiting call is the natural fit here — no event loop to keep unblocked.
    private val waiters = ConcurrentHashMap<String, CopyOnWriteArrayList<CountDownLatch>>()

    override fun onCreate(db: SQLiteDatabase) {
        db.execSQL(
            """
            CREATE TABLE blobs (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                tag TEXT NOT NULL,
                blob TEXT NOT NULL,
                created_at INTEGER NOT NULL
            )
            """.trimIndent()
        )
        db.execSQL("CREATE INDEX idx_blobs_tag ON blobs(tag)")
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        // The queue is a cache, not a record — safe to drop and recreate on a future schema
        // bump rather than migrate, same reasoning as the relay having no read log at all.
        db.execSQL("DROP TABLE IF EXISTS blobs")
        onCreate(db)
    }

    /** Returns false (and stores nothing) if [tag]'s queue is already at capacity. */
    fun put(tag: String, blobBase64: String): Boolean {
        val db = writableDatabase
        val count = DatabaseUtils.longForQuery(db, "SELECT COUNT(*) FROM blobs WHERE tag = ?", arrayOf(tag))
        if (count >= RelayConstants.MAX_BLOBS_PER_TAG) return false
        val totalCount = DatabaseUtils.longForQuery(db, "SELECT COUNT(*) FROM blobs", null)
        if (totalCount >= RelayConstants.MAX_TOTAL_BLOBS) return false
        val values = ContentValues().apply {
            put("tag", tag)
            put("blob", blobBase64)
            put("created_at", System.currentTimeMillis())
        }
        val inserted = db.insert("blobs", null, values) != -1L
        if (inserted) {
            waiters[tag]?.forEach { it.countDown() }
        }
        return inserted
    }

    /**
     * Blocks the calling thread until [tags] has a new blob on any of them, or [timeoutMs]
     * elapses — whichever comes first. The caller must always re-check the actual queues after
     * this returns (see RelayHttpServer's fetchMany handler) rather than trust that something
     * arrived: several tags could have fired at once, and a timeout returns the same way a real
     * arrival does.
     */
    fun waitForAny(tags: List<String>, timeoutMs: Long) {
        if (tags.isEmpty() || timeoutMs <= 0) return
        val latch = CountDownLatch(1)
        for (tag in tags) {
            waiters.getOrPut(tag) { CopyOnWriteArrayList() }.add(latch)
        }
        try {
            latch.await(timeoutMs, TimeUnit.MILLISECONDS)
        } finally {
            for (tag in tags) {
                val list = waiters[tag] ?: continue
                list.remove(latch)
                if (list.isEmpty()) waiters.remove(tag, list)
            }
        }
    }

    /** One picked-up blob plus when it was originally stored — [createdAt] only exists so the
     *  caller can measure dwell time (how long it sat queued); nothing here is kept once this
     *  call returns, matching the class's no-read-log design. */
    data class StoredBlob(val blob: String, val createdAt: Long)

    /** Atomically returns and deletes every blob queued for [tag], oldest first — delete-on-
     *  fetch, no read log, no history. Mirrors blobStore.ts's takeAll transaction exactly. */
    fun takeAll(tag: String): List<StoredBlob> {
        val db = writableDatabase
        db.beginTransaction()
        try {
            val ids = ArrayList<Long>()
            val blobs = ArrayList<StoredBlob>()
            db.rawQuery(
                "SELECT id, blob, created_at FROM blobs WHERE tag = ? ORDER BY id ASC",
                arrayOf(tag)
            ).use { c ->
                while (c.moveToNext()) {
                    ids.add(c.getLong(0))
                    blobs.add(StoredBlob(c.getString(1), c.getLong(2)))
                }
            }
            for (id in ids) db.delete("blobs", "id = ?", arrayOf(id.toString()))
            db.setTransactionSuccessful()
            return blobs
        } finally {
            db.endTransaction()
        }
    }

    /** Deletes blobs older than [ttlMs]. Returns how many were removed. */
    fun sweepExpired(ttlMs: Long, now: Long = System.currentTimeMillis()): Int {
        val cutoff = now - ttlMs
        return writableDatabase.delete("blobs", "created_at < ?", arrayOf(cutoff.toString()))
    }

    /** Manual full reset — the user-triggered counterpart to the automatic TTL sweep. */
    fun wipeAll(): Int = writableDatabase.delete("blobs", null, null)

    fun count(): Long = DatabaseUtils.longForQuery(writableDatabase, "SELECT COUNT(*) FROM blobs", null)
}

File: unpruuf/server/src/store/blobStore.ts
import Database from "better-sqlite3";
import { EventEmitter } from "events";
import { DEFAULT_TTL_HOURS, MAX_BLOBS_PER_TAG } from "../config";
import { getSqliteNativeBinding } from "../sqliteNativeBinding";

/**
 * Blind store-and-forward queue, addressed only by the sender's rotating wire tag (see
 * IdentityManager.myWireId in the Android app). The relay never learns contact identity or
 * message content beyond an already Double-Ratchet-encrypted, opaque blob — same pseudonymity
 * model as the direct P2P path, just with a mailbox instead of a live socket.
 */
export class BlobStore {
  private db: Database.Database;
  private ttlMs: number;
  // Backs /v1/fetchMany's optional long-poll (see routes/relay.ts) — a client can wait on
  // several tags at once instead of blindly re-polling every few seconds. One process-local
  // emitter, not persisted: a missed event just means the next regular poll (or the next
  // fetchMany call's own immediate check, which always runs regardless of this) still finds it,
  // since this is purely a "wake up sooner" optimization, never the only way data is found.
  // Unbounded listener count on purpose — one listener per concurrently-waiting fetchMany call
  // per tag, and a busy relay can legitimately have many of both.
  private events = new EventEmitter().setMaxListeners(0);

  /** [ttlHours] is operator-configurable — see identity.ts; defaults to DEFAULT_TTL_HOURS. */
  constructor(dbPath: string, ttlHours: number = DEFAULT_TTL_HOURS) {
    this.ttlMs = ttlHours * 60 * 60 * 1000;
    const nativeBinding = getSqliteNativeBinding();
    // @types/better-sqlite3 types `Options.nativeBinding` as `string` only, but the actual
    // runtime (lib/database.js) also accepts an already-loaded addon object — confirmed by
    // reading that file directly, not assumed. The types just haven't caught up.
    const options = nativeBinding ? ({ nativeBinding } as unknown as Database.Options) : undefined;
    this.db = new Database(dbPath, options);
    this.db.pragma("journal_mode = WAL");
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS blobs (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        tag TEXT NOT NULL,
        blob TEXT NOT NULL,
        created_at INTEGER NOT NULL
      );
      CREATE INDEX IF NOT EXISTS idx_blobs_tag ON blobs(tag);
    `);
  }

  /** Returns false (and stores nothing) if the tag's queue is already at capacity. */
  put(tag: string, blobBase64: string): boolean {
    const count = this.db.prepare("SELECT COUNT(*) as n FROM blobs WHERE tag = ?").get(tag) as { n: number };
    if (count.n >= MAX_BLOBS_PER_TAG) return false;
    this.db
      .prepare("INSERT INTO blobs (tag, blob, created_at) VALUES (?, ?, ?)")
      .run(tag, blobBase64, Date.now());
    this.events.emit(`tag:${tag}`);
    return true;
  }

  /**
   * Resolves as soon as ANY of [tags] receives a new blob, or after [timeoutMs] — whichever
   * comes first. Never rejects. The caller must always re-check the actual queues after this
   * resolves (see routes/relay.ts) rather than trust which tag fired: several could have arrived
   * at once, and a timeout resolves the same way a real arrival does.
   */
  waitForAny(tags: string[], timeoutMs: number): Promise<void> {
    if (tags.length === 0 || timeoutMs <= 0) return Promise.resolve();
    const events = this.events;
    return new Promise((resolve) => {
      let done = false;
      let timer: ReturnType<typeof setTimeout>;
      const finish = () => {
        if (done) return;
        done = true;
        clearTimeout(timer);
        for (const { tag, listener } of listeners) events.off(`tag:${tag}`, listener);
        resolve();
      };
      const listeners = tags.map((tag) => {
        const listener = () => finish();
        events.once(`tag:${tag}`, listener);
        return { tag, listener };
      });
      timer = setTimeout(finish, timeoutMs);
    });
  }

  /** Atomically returns and deletes every blob queued for [tag], oldest first. */
  takeAll(tag: string): string[] {
    return this.takeAllWithMeta(tag).map((r) => r.blob);
  }

  /**
   * Same atomic take-and-delete as [takeAll], but also returns each blob's `createdAt` so the
   * caller can report how long it sat queued before pickup (see `relayEventLog.ts` — that dwell
   * time is the single most useful number for an operator diagnosing "are messages arriving but
   * not being collected, or not arriving at all?"). [takeAll] delegates here so there is only
   * one copy of the transaction logic.
   */
  takeAllWithMeta(tag: string): { blob: string; createdAt: number }[] {
    const take = this.db.transaction((t: string) => {
      const rows = this.db
        .prepare("SELECT id, blob, created_at FROM blobs WHERE tag = ? ORDER BY id ASC")
        .all(t) as { id: number; blob: string; created_at: number }[];
      if (rows.length > 0) {
        const del = this.db.prepare("DELETE FROM blobs WHERE id = ?");
        for (const row of rows) del.run(row.id);
      }
      return rows.map((r) => ({ blob: r.blob, createdAt: r.created_at }));
    });
    return take(tag);
  }

  /**
   * Operator-facing snapshot of what is currently waiting to be collected — backs the `--status`
   * command and the running relay's periodic heartbeat. Deliberately aggregate-only: counts and
   * the age of the oldest item, never tags or blobs, so reading it stays consistent with the
   * relay's blindness to who is talking to whom.
   */
  stats(now: number = Date.now()): { queued: number; tags: number; oldestAgeMs: number | null } {
    const row = this.db
      .prepare("SELECT COUNT(*) AS queued, COUNT(DISTINCT tag) AS tags, MIN(created_at) AS oldest FROM blobs")
      .get() as { queued: number; tags: number; oldest: number | null };
    return {
      queued: row.queued,
      tags: row.tags,
      oldestAgeMs: row.oldest === null ? null : now - row.oldest,
    };
  }

  /** Deletes blobs older than this store's configured TTL. Returns how many were removed. */
  sweepExpired(now: number = Date.now()): number {
    const cutoff = now - this.ttlMs;
    const result = this.db.prepare("DELETE FROM blobs WHERE created_at < ?").run(cutoff);
    return result.changes;
  }

  close(): void {
    this.db.close();
  }
}
